### Contents


### 1.0 Import Libraries

### 2.0 Import Data

### 3.0 Consistency Checks

### 4.0 Task Questions

# 1.0 Import libraries

In [1]:
#import libraries
import pandas as pd
import numpy as np
import os

# 2.0 import data

In [3]:
#define path to project folder
path = r'C:\Users\Asus 17\OneDrive\Desktop\Achievement 4\instacart_basket_analysis'

In [4]:
# import products.csv as df_prods
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [54]:
# import orders_wrangled.csv as df_ords
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'))

## Check dfs


In [55]:
# check prods looks ok
df_prods.head(10)

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   
5           6                                       Dry Nose Oil        11   
6           7                     Pure Coconut Water With Orange        98   
7           8                  Cut Russet Potatoes Steam N' Mash       116   
8           9                  Light Strawberry Blueberry Yogurt       120   
9          10     Sparkling Orange Juice & Prickly Pear Beverage       115   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3  
5             11     2.6  
6              7     4.4  
7              1     1.1  
8             16     7.0  
9              7     8.4

In [56]:
df_prods.tail(10)

product_id                                product_name  aisle_id  \
49683       49679                     Famous Chocolate Wafers        61   
49684       49680          All Natural Creamy Caesar Dressing        89   
49685       49681     Spaghetti with Meatballs and Sauce Meal        38   
49686       49682                          California Limeade        98   
49687       49683                              Cucumber Kirby        83   
49688       49684   Vodka, Triple Distilled, Twist of Vanilla       124   
49689       49685          En Croute Roast Hazelnut Cranberry        42   
49690       49686                            Artisan Baguette       112   
49691       49687  Smartblend Healthy Metabolism Dry Cat Food        41   
49692       49688                      Fresh Foaming Cleanser        73   

       department_id  prices  
49683             19     6.0  
49684             13     4.9  
49685              1     6.9  
49686              7     4.3  
49687              4    13.2  
49688              5     5.3  
49689              1     3.1  
49690              3     7.8  
49691              8     4.7  
49692             11    13.5

In [57]:
#check ords
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
0           0   2539329        1             1                  2   
1           1   2398795        1             2                  3   
2           2    473747        1             3                  3   
3           3   2254736        1             4                  4   
4           4    431534        1             5                  4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [58]:
df_ords.tail()

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
3421078     3421078   2266710   206209            10                  5   
3421079     3421079   1854736   206209            11                  4   
3421080     3421080    626363   206209            12                  1   
3421081     3421081   2977660   206209            13                  1   
3421082     3421082    272231   206209            14                  6   

         order_hour_of_day  days_since_prior_order  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0

# 3.0 Consistency Checks

In [59]:
# describe orders
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       order_day_of_week  order_hour_of_day  days_since_prior_order  
count       3.421083e+06       3.421083e+06            3.214874e+06  
mean        2.776219e+00       1.345202e+01            1.111484e+01  
std         2.046829e+00       4.226088e+00            9.206737e+00  
min         0.000000e+00       0.000000e+00            0.000000e+00  
25%         1.000000e+00       1.000000e+01            4.000000e+00  
50%         3.000000e+00       1.300000e+01            7.000000e+00  
75%         5.000000e+00       1.600000e+01            1.500000e+01  
max         6.000000e+00       2.300000e+01            3.000000e+01

In [60]:
# describe prods
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

## running a mixed type test

In [61]:
# Create a dataframe
df_test = pd.DataFrame()

In [62]:
# create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [63]:
# check that it worked
df_test.head()

mix
0     a
1     b
2     1
3  True

In [64]:
# check for mixed types
for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1) 
    if len (df_test[weird]) > 0:
        print (col)

mix


In [65]:
#create a numeric column for the lols
df_test['numeric'] = [1, 2, 3, 4]

In [66]:
df_test.head()

mix  numeric
0     a        1
1     b        2
2     1        3
3  True        4

In [67]:
# check for mixed types again
for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_test[weird]) > 0:
        print (col)

mix


In [68]:
# change mix to strings
df_test['mix'] = df_test['mix'].astype('str')

In [69]:
#check again 
for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_test[weird]) > 0:
        print (col)

## Finding Missing Values

In [70]:
# looking for null values in prods
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [71]:
# seeing what happens without .sum()
df_prods['product_name'].isnull()

0        False
1        False
2        False
3        False
4        False
         ...  
49688    False
49689    False
49690    False
49691    False
49692    False
Name: product_name, Length: 49693, dtype: bool

In [72]:
# make a subset of NA values
df_nan = df_prods[df_prods['product_name'].isnull() == True] 

In [73]:
# check it out
df_nan.head(20)

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [74]:
# how big is df_prods?
df_prods.shape

(49693, 5)

In [75]:
#make a new df_prods without the nan values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [76]:
# have a look
df_prods_clean.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [77]:
# has the shape changed by 16 rows?
df_prods_clean.shape

(49677, 5)

In [78]:
49693 - 49677 

16

narator: it had

## looking for duplicates in prods


In [79]:
# finding df_prods_clean for duplicates and making them their own df
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [80]:
#taking a look
df_dups.head()

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [81]:
#checking the size
df_dups.shape

(5, 5)

## dropping the duplicates


In [82]:
# check original row count
df_prods_clean.shape

(49677, 5)

In [83]:
# create a new duplicate-free dataframe
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [84]:
# check no_dups row count
df_prods_clean_no_dups.shape

(49672, 5)

In [85]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

# 4.0  Task Questions. 
## Q 2 


In [86]:
# having a look at df_ords
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       order_day_of_week  order_hour_of_day  days_since_prior_order  
count       3.421083e+06       3.421083e+06            3.214874e+06  
mean        2.776219e+00       1.345202e+01            1.111484e+01  
std         2.046829e+00       4.226088e+00            9.206737e+00  
min         0.000000e+00       0.000000e+00            0.000000e+00  
25%         1.000000e+00       1.000000e+01            4.000000e+00  
50%         3.000000e+00       1.300000e+01            7.000000e+00  
75%         5.000000e+00       1.600000e+01            1.500000e+01  
max         6.000000e+00       2.300000e+01            3.000000e+01

Ok, I've got a funny unnamed column that seems to be an index. 
I'm interested in if order numbers per customer are somehow capped at 100. That seems a little neat. 

## Question 3


In [93]:
#check for mixed-type data in ords
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
        print (col)

In [95]:
#check data types for ords
df_ords.dtypes

Unnamed: 0                  int64
order_id                    int64
user_id                     int64
order_number                int64
order_day_of_week           int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

No Mixed-type columns were found. 


## question 5

In [100]:
#looking for total of null values in ords
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
order_day_of_week              0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

206209 null values in days since prior order makes sense because according to the data dictionary this column is null on first order


## Question 6

In [101]:
# make a subset dataframe of the rows with null values
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [102]:
#check it out
df_ords_nan.head()

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
0            0   2539329        1             1                  2   
11          11   2168274        2             1                  2   
26          26   1374495        3             1                  1   
39          39   3343014        4             1                  6   
45          45   2717275        5             1                  3   

    order_hour_of_day  days_since_prior_order  
0                   8                     NaN  
11                 11                     NaN  
26                 14                     NaN  
39                 11                     NaN  
45                 12                     NaN

In [103]:
#check that they are all on order_number == 1
df_ords_nan['order_number'].value_counts(dropna = False)

order_number
1    206209
Name: count, dtype: int64

These are important and valid readings and don't need to be changed. It would significantly alter our Data if we were to delete every customer's first order, or to impute a a central value here. 




## Question 7

In [106]:
# create a subset for duplicates in ords
df_ords_dups = df_ords[df_ords.duplicated()]

In [107]:
# have a look at the subset
df_ords_dups.head()

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, order_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

No duplicates found

# export dataframes


In [110]:
# export df_prods to products_checked.csv
df_prods.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))

In [112]:
# export df_ords to orders_checked.csv
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))